[blog] Add `Grid v2` announcement (#33926)
mui/material-ui: docs/pages/blog/hello-grid-version-2.md

---

> title: Hello, Grid version 2

Reviewer: What could we name this piece to try to hook the reader? What could we tell a developer reading our blog or social media feed to convince them that they should click and check out this article?

Author: - "Things you should know about the new grid component"
- "Here is how the new grid makes your life easier"
- "Build layout faster with the new grid component"

Any of these catch your attention?

Reviewer: I like the last one best:

```suggestion
title: Build layouts faster with the new Grid component
```

description: We're excited to introduce the Grid version 2 that comes with new features and improvements!
date: 2022-08-20T00:00:00.000Z
authors: ['siriwatknp']
tags: ['MUI Core', 'News']
card: false
---

> We have shipped the new `Grid` component in [Material UI v5.9.0](https://github.com/mui/material-ui/releases/tag/v5.9.0) that will replace the existing one in the next major release.

Author: Much better!

```js
import Grid from '@mui/material/Grid'; // The current grid, a.k.a. Grid v1

> import Grid2 from '@mui/material/Unstable_Grid2'; // The new grid, a.k.a. Grid v2

Author: The import name depends on the developers but we could make it clear in the demos to use `Grid2`.

The snippets below produce the same result.

```js
import Grid2 from '@mui/material/Unstable_Grid2';

// the usage should be Grid2
<Grid2>...</Grid2>
```

```js
import Grid from '@mui/material/Unstable_Grid2';

// the usage should be Grid
<Grid>...</Grid>
```

> import Grid2 from '@mui/material/Unstable_Grid2'; // The new grid, a.k.a. Grid v2

Reviewer: I see. We should be consistent throughout the doc, so either we call it `Grid` instead of `Grid2` in this `import`, or else we should change all instances of `<Grid>` to `<Grid2>` in the examples. I think we should just call it `Grid` everywhere for the sake of simplicity.

> import Grid2 from '@mui/material/Unstable_Grid2'; // The new grid, a.k.a. Grid v2

Author: will go with `Grid`.

```

## The problems and motivation

- To fix the migration pain points ([#26266](https://github.com/mui/material-ui/issues/29266) and [#31244](https://github.com/mui/material-ui/issues/31244)) for the grid component which is caused by the [implementation changes](https://github.com/mui/material-ui/pull/24332) introduced in v5.
- It is an opportunity to add the long-awaited [offset feature](https://github.com/mui/material-ui/issues/11251) to the grid component.
- The grid container is using child combinator (`>`) to style the grid items which makes [customization](https://github.com/mui/material-ui/issues/28855) harder.
- The [limitation](https://mui.com/material-ui/react-grid/#nested-grid) on nested grids could be fixed.
- Unnecessary `item` prop.

## What's new

> The `Grid v2` has been rewritten from scratch with CSS variables and `calc()` which lets us fix the issues and apply new features easily.

Author: Got it.

### 💥 Breaking change

The negative margin in `Grid v2` spreads equally on all sides by default (same as the `Grid` in Material UI v4).

:::info

> We believe that the migration from Grid `v1` to `v2` will be smooth for most of the cases.

Author: Will keep this in mind 👍

### ✨ Improvements

#### Simplifying logic

Some parts of the calculation are delegated to CSS by using CSS variables and `calc()` together. They reduces the need of javascript and makes the component scalable.

For example, the `Grid v1` currently calculates the width by reading the value from a React context. However, with CSS variables in place, we can remove the React context and hand over the work to CSS instead:

```js
{
  // --Grid-columns is defined in the grid container
  width: `calc(100% * ${value} / var(--Grid-columns))`,
}
```

Another example is how the padding is calculated in grid items:

In `Grid v1`, grid items has no information about the padding it should have because the `spacing` prop is defined on the grid container. So, the simplest way is to control the padding from the grid container using CSS child combinator (`>`):

```js
function getOffset(val) {
  const parse = parseFloat(val);
  return `${parse}${String(val).replace(String(parse), '') || 'px'}`;
}
// ...The code is shorten for readability
({
  [`& > .${gridClasses.item}`]: {
    // Grid item's padding is controlled by its container
    paddingTop: getOffset(themeSpacing),
  },
});
```

> However, in `Grid v2`, the javascript logic can be separated independently because CSS variables shift the communication between container and item to CSS.

Reviewer: ```suggestion
In v2, the JavaScript logic is independent from the CSS, because CSS variables shift the communication between container and item over to the CSS.
```

Author: > the JavaScript logic is independent from the CSS

This phrase has a different meaning from what I intended to communicate.

Or should it be
```suggestion
In v2, the JavaScript logic is independent between grid container and item, because CSS variables shift the communication over to the CSS.
```

Reviewer: Ohhh I see what you're saying, something like this maybe?

```suggestion
In v2, the communication between the grid container and its items happens via CSS instead of JavaScript, so that the logic behind these components can be better separated.
```

> - **Grid container** takes care of the variables generation based on the `spacing` prop:

Reviewer: ```suggestion
- The grid container takes care of generating variables based on the `spacing` prop:
```

  ```js
  // The code is simplified for readability
  {
    '--Grid-rowSpacing': spacingProp,
    '--Grid-columnSpacing': spacingProp,
  }
  ```

> - **Grid item** refers to the variables:

Reviewer: ```suggestion
- The grid item then makes references to those variables:
```

  ```js
  {
    padding: `calc(var(--Grid-rowSpacing) / 2) calc(var(--Grid-columnSpacing) / 2)`,
  }
  ```

> This is a very good use case for applying CSS variables when it comes to components that has a parent-child relationship which also support media query customization.

Reviewer: ```suggestion
This is a very good use case for applying CSS variables when it comes to components that have a parent-child relationship and also support media query customization.
```

#### Remove unnecessary props

> In version 2, `Grid` is always an item which is similar to the flexbox item in CSS, so the `item` prop is not needed anymore.

Reviewer: ```suggestion
In version 2, `Grid` is always an item—similar to the Flexbox item in CSS—so the `item` prop is no longer needed.
```

> We also update the grid styles to work with ellipsis text by default, so the `zeroMinWidth` prop is removed.

Reviewer: ```suggestion
We've also updated the grid styles to work with ellipsis text by default, so the `zeroMinWidth` prop has been removed:
```

Reviewer: What exactly is meant by "work with ellipsis text"? Does that mean it supports text that includes an ellipsis?

Author: My intention is to say that developers do not need to use `zeroMinWidth` together with `<Typography nowrap>` anymore.

https://mui.com/material-ui/react-grid/#white-space-nowrap

Reviewer: Ah ok, that's a helpful example. I think I would describe it like this (if I'm understanding correctly):

```suggestion
The new `Grid` automatically handles the sizing of truncated text that's too long for its container, so we've also removed the `zeroMinWidth` prop since it's no longer needed:
```

Author: awesome!

```diff
 <Grid container spacing={2}>
-  <Grid item zeroMinWidth><Typography>Long text...</Typography></Grid>
+  <Grid><Typography>Long text...</Typography></Grid>
 </Grid>
```

#### Nested grid

> With CSS variables and the removal of the `item` prop, there is no more limitation for creating nested grids.

Reviewer: ```suggestion
With the addition of CSS variables and the removal of the `item` prop, there are no more limitations when it comes to creating nested grids.
```

> As a bonus, a grid container automatically inherits row and column spacing values from the root grid container unless they are specified directly on the component.

Reviewer: ```suggestion
As a bonus, a grid container automatically inherits row and column spacing values from the root grid container, unless they are specified directly on the component.
```

```js
import Grid from '@mui/material/Unstable_Grid2`;

// root grid container
<Grid container spacing={2}>
  <Grid>...</Grid>
  <Grid container> {/* inherits spacing from the root container */}
    <Grid>...</Grid>
    <Grid>...</Grid>
  </Grid>
  <Grid>...</Grid>
</Grid>
```

## Future plan and migration

> Since the `Grid v2` is rewritten from scratch, it is currently considered Unstable\_ as we give the community time to try it out and offer feedback. We will make it stable and deprecate v1 in the next major release of Material UI.

Reviewer: ```suggestion
Since the `Grid` has been rewritten from scratch for v2, it is currently considered _unstable_ as we give the community time to try it out and offer feedback. 
We will make it stable and deprecate v1 in the next major release of Material UI.
```

> Are you convinced? check out the [demos](/material-ui/react-grid2/) and the [migrating to Grid v2](/material-ui/migration/migration-grid-v2/) pages.

Reviewer: ```suggestion
Ready to make the jump? 
Check out the [Grid v2 documentation](/material-ui/react-grid2/) and the [Grid v2 migration guide](/material-ui/migration/migration-grid-v2/).
```

<hr />

> **That's it for today!** Happy coding 👨‍💻👩‍💻!.

Reviewer: ```suggestion
That's it for today! Happy coding 👨‍💻👩‍💻!.
```

---
title: Hello, Grid version 2

> description: We're excited to introduce the Grid version 2 that comes with new features and improvements!

Reviewer: Instead of saying it comes with new features, let's mention some of them directly.

```suggestion
description: The new Grid features simplified logic, support for offsetting and nested grids, and more.
```
